aurora-is-near/rainbow-bridge
# Upgrade to nearcore 1.23.0

## Motivation
Close issue #644 (check issue for details)

## Changes

 `NearDecoder`: upgrade to `nearcore` 1.23.0:
* Handle enum byte for `ValidatorStake` version.
* `BlockProducer`: add `isChunkOnly` field.

Near2Eth-relay:
* utils/borsh.js: add support for optional `validator_stake_struct_version` field
  when borshifying light client block.

## Additional resources

You can check the [SSCCE](https://github.com/sept-en/borsh-validators-example/blob/master/src/lib.rs) verifying the change of the borsh-serialized result after wrapping the structure into an enum. 

## Tests

There are some tests that are failing due to this upgrade. Those tests rely on the old blocks which don't contain the enum for `ValidatorStake` version thus their hash will differ from the hash that will be computed having new rules with versioning (nearcore v1.23.0 upgrade adds an extra byte as a prefix to block producer which changes the hash). We can't change only `next_bp_hash` in files containing the blocks for tests as in this case signatures of validators will mismatch. The possible solutions to resolve tests:

1) Add another contract `NearBridgeV1` that will contain the old schema for decoding blocks prior to nearcore v1.23.0.
2) Add the possibility to compute the signature for older blocks correctly in `NearDecoder`.
2) Wait until the nearcore v1.23.0 is live on Testnet and after that use real Testnet blocks to update all failing tests.

**Option 1:** IMO is not really useful as we will test against old code-base that will not be supported anymore. Still, this will give us some insights if some tool is broken (e.g. near2eth-relay). But nevertheless, new tests for after-upgrade blocks should be added. So this looks like a redundant step.
**Option 2:** this will add another branch, increase the complexity of the code and this means that we are supporting an obsolete code. Also, this will include additional gas costs because the signature won't be computed from the single blob anymore.

So having all these things in mind, I'm for the **3rd** option but would like to hear any suggestions/ideas.
